src/hotspot/cpu/x86/x86_64.ad

```diff
@@ -7574,14 +7682,70 @@ instruct decL_mem(memory dst, immL_M1 src, rFlagsReg cr)
   ins_pipe(ialu_mem_imm);
 %}
 
-instruct leaL_rReg_immL(rRegL dst, rRegL src0, immL32 src1)
+instruct leaL_rReg_immI2_immL32(rRegL dst, rRegL index, immI2 scale, immL32 disp)
 %{
-  match(Set dst (AddL src0 src1));
+  predicate(VM_Version::supports_fast_2op_lea());
+  match(Set dst (AddL (LShiftL index scale) disp));
 
-  ins_cost(110);
-  format %{ "leaq    $dst, [$src0 + $src1]\t# long" %}
+  format %{ "leaq $dst, [$index << $scale + $disp]\t# long" %}
   ins_encode %{
-    __ leaq($dst$$Register, Address($src0$$Register, $src1$$constant));
+    Address::ScaleFactor scale = static_cast<Address::ScaleFactor>($scale$$constant);
+    __ leaq($dst$$Register, Address(noreg, $index$$Register, scale, $disp$$constant));
+  %}
+  ins_pipe(ialu_reg_reg);
+%}
+
+instruct leaL_rReg_rReg_immL32(rRegL dst, rRegL src1, rRegL src2, immL32 disp)
+%{
+  predicate(VM_Version::supports_fast_3op_lea());
+  match(Set dst (AddL (AddL src1 src2) disp));
+
+  format %{ "leaq $dst, [$src1 + $src2 + $disp]\t# long" %}
+  ins_encode %{
+    __ leaq($dst$$Register, Address($src1$$Register, $src2$$Register, Address::times_1, $disp$$constant));
+  %}
+  ins_pipe(ialu_reg_reg);
+%}
+
+instruct leaL_rReg_rReg_immI2(rRegL dst, rRegL base, rRegL index, immI2 scale)
+%{
+  predicate(VM_Version::supports_fast_3op_lea());
+  match(Set dst (AddL base (LShiftL index scale)));
+  match(Set dst (AddL (LShiftL index scale) base));
+
+  format %{ "leaq $dst, [$base + $index << $scale]\t# long" %}
+  ins_encode %{
+    Address::ScaleFactor scale = static_cast<Address::ScaleFactor>($scale$$constant);
+    __ leaq($dst$$Register, Address($base$$Register, $index$$Register, scale));
+  %}
+  ins_pipe(ialu_reg_reg);
+%}
+
+instruct leaL_rReg_rReg_immI2_no_disp(rRegL dst, no_rbp_r13_RegL base, rRegL index, immI2 scale)
```

Comment:
Does it make too much sense to keep both `leaL_rReg_rReg_immI2_no_disp ` and `leaL_rReg_rReg_immI2 ` for real programs?

The more instruct rules in the AD file, the bigger of the libjvm.so would be.
I would suggest removing `leaL_rReg_rReg_immI2 `.

Reply:
I agree